api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java

```diff
@@ -816,6 +836,9 @@ public void execute() {
 
     @Override
     public void create() throws ResourceAllocationException {
+        if (!isVolumeOrSnapshotProvided() && templateId == null) {
+            throw new CloudRuntimeException("There isn't a ROOT volume, snapshot of ROOT volume or template provided to deploy a Virtual machine");
+        }
```

Comment:
maybe good to extract a validate method and also check for double parameters; i.e. it is amniguous if two or all three of the parameters are given.

Reply:
+1 a checkParameters method would be necessary to cover all the new supported cases + the existing one (only templateId set)